### ISSUE TITLE
add command type based on a query

### ISSUE THREAD

**Issue by jakildev** [jakildev/IrredenEngine#17]
kinda like running a system tick function one time

**Comment by jakildev** [comment]
## Plan: add a one-shot query-command primitive (query-based command execution)

- **Issue:** #17
- **Model:** sonnet — the API design is fully settled below; implementation is mechanical (inline templates mirroring existing `detail::` machinery, switch-case wiring, tests from existing fixtures)
- **Date:** 2026-07-03

*(Planned during the 2026-07-03 backlog triage by the architect under human delegation — files-with-plan path, no worker re-plan needed.)*

### Scope

Add a run-once query-execution primitive to `IRSystem` — the run-now counterpart to `createSystemDynamic`: same archetype-node traversal, no `SystemId`, no pipeline registration, nothing persists. Then implement `Command<RANDOMIZE_VOXELS>` on top as the first consumer, honoring the enum's "exclude locked entities" note via a new `C_Locked` tag component. Lua exposure of the primitive itself is out of scope (noted below); `RANDOMIZE_VOXELS` becomes Lua-reachable for free through the existing `fireByName` binding.

### Verified current state

- `RANDOMIZE_VOXELS` enum entry + same-era TODO ("should operate like a system with a query... exclude locked entities") — `engine/command/include/irreden/command/ir_command_types.hpp:38-39`.
- `engine/prefabs/irreden/voxel/commands/command_randomize_voxels.hpp` is an empty header (include guard only) — no `Command<RANDOMIZE_VOXELS>` specialization exists.
- `engine/command/src/ir_command.cpp`: `RANDOMIZE_VOXELS` sits in the no-specialization fall-through of BOTH `bindPrefabCommand` (line 235) and `fireByName` (line 330); firing it today logs "no Command<> specialization registered" and no-ops.
- `SystemManager::createSystemDynamic` — declared `engine/system/include/irreden/system/system_manager.hpp:168`, implemented `engine/system/src/system_manager.cpp:48`. It registers persistent state (`m_ticks`, `m_beginTicks`, timing, concurrency slots) and asserts against `PARALLEL_FOR`. There is no run-once primitive.
- The traversal is already factored: `executeSystem` queries nodes via `IREntity::queryArchetypeNodesSimple(include, exclude)` (`engine/system/src/system_manager.cpp:426`; declared `engine/entity/include/irreden/ir_entity.hpp:58`) and dispatches per node. `IREntity::forEachComponent` (ir_entity.hpp:112) is the nearest one-shot sibling but is single-component with no exclude support.
- Facade conventions: `IRSystem::createSystemDynamic` free-function wrapper at `engine/system/include/irreden/ir_system.hpp:393`; `Exclude<...>` at `ir_system_types.hpp:252` with `detail::PartitionExcludes` / `ArchetypeFromList` / `FilterTags_t` helpers in ir_system.hpp.
- Already wired (no change needed): `commandNameToString` case (`ir_command.hpp:109-110`) and `IR_BIND_CMD(RANDOMIZE_VOXELS)` (`engine/script/include/irreden/script/lua_command_bindings.hpp:67`).
- No locked-tag component exists anywhere in `engine/prefabs/`; `C_Persistent` (`component_persistent.hpp`) is the empty-tag precedent.
- `C_VoxelSetNew::editVoxels(fn(int, C_Voxel&, vec3))` (`component_voxel_set.hpp:485`) is the sanctioned bulk-edit entry point (#2165) and is headless-testable via the explicit `targetCanvas` ctor (fixture in `test/ecs/voxel_set_edit_api_test.cpp:63-70`). `IRMath::randomColor()` returns alpha 255 (`engine/math/src/ir_math.cpp:44`).
- Test target `IrredenEngineTest` exists with an `ecs/` group (`test/CMakeLists.txt:20-39`); `test/ecs/system_exclude_test.cpp` is the exact fixture template (bare `EntityManager` + `SystemManager`, TU-local components).

### Approach

Single committed approach: one-shot query surface in `IRSystem` (matches the issue's "running a system tick function one time" mental model), traversal shared via `IREntity::queryArchetypeNodesSimple` — the same call `executeSystem` makes, so nothing is duplicated.

1. **`engine/system/include/irreden/ir_system.hpp`** — add two inline free functions after the `createSystemDynamic` wrapper (~line 407), following the `execute*` = run-now verb family (`executePipeline`, `executeSystem`):
   - `template <typename PerNodeFn> void executeQueryDynamic(const IREntity::Archetype& includeArchetype, const IREntity::Archetype& excludeArchetype, PerNodeFn&& body)` — runtime-typed core, mirrors `createSystemDynamic`'s per-matched-`ArchetypeNode*` body contract. Implementation: `IR_ASSERT_MAIN_THREAD()` (include `irreden/system/ir_assert_main_thread.hpp`), then `queryArchetypeNodesSimple` + `body(node)` per node. No SystemManager touch, no name, no timing slot.
   - `template <typename... QueryComponents, typename FunctionTick> void executeQuery(FunctionTick&& tick)` — compile-time-typed wrapper accepting the same `Exclude<...>` markers as `createSystem` (reuse `detail::PartitionExcludes` for the exclude archetype, `detail::FilterTags_t` + `detail::ArchetypeFromList` for the include). Delegates to `executeQueryDynamic` with a per-node lambda that resolves columns once via `getComponentData<Cs>(node)` and row-iterates, dispatching `tick(Cs&...)` or `tick(EntityId, Cs&...)` via `std::is_invocable_v` (add a small `detail::` TypeList-expander struct beside `CallCreateSystem`). Batch/whole-node consumers use `executeQueryDynamic` directly — the two row forms are enough for commands.
   - Threading stance: **caller thread, serial, main-thread-only** — same rationale as `createSystemDynamic`'s `PARALLEL_FOR` assert (no ranged binder; `getComponentData`'s manager hash lookups aren't worker-safe). Doc-comment the deferred-ops contract: bodies must use `IREntity::deferred*` for structural changes; `executeQuery` does not flush (the pipeline's group boundary does).
2. **`engine/prefabs/irreden/common/components/component_locked.hpp`** (new) — empty tag `struct C_Locked {};` mirroring `C_Persistent`, doc'd as "excluded from bulk query-driven mutation commands (#17)".
3. **`engine/prefabs/irreden/voxel/commands/command_randomize_voxels.hpp`** — implement `Command<RANDOMIZE_VOXELS>::create()` returning a lambda that runs `IRSystem::executeQuery<C_VoxelSetNew, IRSystem::Exclude<C_Locked>>(...)`; per set, `set.editVoxels(...)`: skip voxels with `color_.alpha_ == 0`, otherwise assign `IRMath::randomColor()` **preserving the original alpha**.
4. **`engine/command/src/ir_command.cpp`** — add the header to the include block (alphabetical); promote `RANDOMIZE_VOXELS` out of both fall-throughs into real cases in `bindPrefabCommand` and `fireByName`.
5. **`engine/command/CMakeLists.txt`** — add `IrredenEngineSystem` to the existing PRIVATE link block (same rationale comment as the existing Rendering/Video/Window entries).
6. **`engine/command/include/irreden/command/ir_command_types.hpp:38-39`** — replace the 2023 TODO comment with a pointer to the implementation.
7. **`test/ecs/execute_query_test.cpp`** (new, modeled on `system_exclude_test.cpp`) + **`test/ecs/command_randomize_voxels_test.cpp`** (new, modeled on `voxel_set_edit_api_test.cpp`'s headless pool/targetCanvas fixture); register both in `test/CMakeLists.txt`.
8. **Docs** — short "One-shot queries (`executeQuery`)" paragraph in `engine/system/CLAUDE.md` under the `createSystemDynamic` section; one-line update to `engine/command/CLAUDE.md`'s command pattern section noting the query-command shape.

Lua note (not in scope): `RANDOMIZE_VOXELS` is immediately reachable from Lua via existing `IRCommand.fireByName` / `IRCommand.bindPrefab` once step 4 lands — zero new binding code. A Lua-facing `IRSystem.executeQuery` one-shot would be a small follow-up in `engine/script/` if ever needed (persistent Lua systems already exist via `IRSystem.registerSystem`).

### Affected files
- `engine/system/include/irreden/ir_system.hpp` — add `executeQueryDynamic` + `executeQuery<Cs...>` + `detail::` row-dispatch helper
- `engine/prefabs/irreden/common/components/component_locked.hpp` — new `C_Locked` empty tag component
- `engine/prefabs/irreden/voxel/commands/command_randomize_voxels.hpp` — implement `Command<RANDOMIZE_VOXELS>` on `executeQuery`, excluding `C_Locked`
- `engine/command/src/ir_command.cpp` — include new header; real cases for `RANDOMIZE_VOXELS` in `bindPrefabCommand` + `fireByName`
- `engine/command/CMakeLists.txt` — PRIVATE-link `IrredenEngineSystem`
- `engine/command/include/irreden/command/ir_command_types.hpp` — resolve the 2023 TODO comment
- `test/ecs/execute_query_test.cpp` — new primitive tests
- `test/ecs/command_randomize_voxels_test.cpp` — new command tests
- `test/CMakeLists.txt` — register the two new test files
- `engine/system/CLAUDE.md`, `engine/command/CLAUDE.md` — doc the primitive + the query-command pattern

### Acceptance criteria
- `fleet-build --target IrredenEngineTest` green; `fleet-run IrredenEngineTest --gtest_filter='ExecuteQuery*:RandomizeVoxels*'` all pass.
- `execute_query_test.cpp` asserts: (a) `executeQuery<Counter>` visits entities across multiple archetypes; (b) `executeQuery<Counter, Exclude<LockTag>>` skips tagged entities; (c) the `EntityId`-form tick receives correct ids; (d) `executeQueryDynamic` fires exactly once per matched node; (e) **no registration**: `SystemManager::getSystemCount()` is unchanged by a call, and a subsequent `executePipeline(UPDATE)` does not re-run the body.
- `command_randomize_voxels_test.cpp` (headless `C_VoxelPool` + explicit-`targetCanvas` sets): two sets with a known uniform color, one entity also carrying `C_Locked`; one carved (alpha-0) voxel in the unlocked set. After `IRCommand::fireByName(IRCommand::RANDOMIZE_VOXELS)`: unlocked set has ≥1 RGB byte changed and every alpha byte preserved (no exact-color asserts — RNG is unseeded); carved voxel still inactive and pool active-mask unchanged; locked set byte-identical.
- Firing `RANDOMIZE_VOXELS` no longer hits the `IRE_LOG_ERROR` unimplemented path (covered by the fireByName-driven test).
- Existing tests still green (notably `system_exclude_test`, `voxel_set_edit_api_test`, `script/lua_command_test`).

### Gotchas
- **`randomColor()` returns alpha 255** — naively assigning it resurrects carved voxels and desyncs the pool active-mask. Preserve the original alpha per voxel; skipping alpha-0 voxels keeps the mask stable through `editVoxels`'s resync.
- **Use `editVoxels`, never raw `voxels_[i]` writes** — the resync also maintains the GRID-rotation source mirror; raw span writes are invalid once a set has begun rotating (#2165 contract, enforced by `simplify-check-ecs`).
- **Edit BOTH switches in ir_command.cpp** — leaving `RANDOMIZE_VOXELS` in a fall-through while adding the real case is a duplicate-case compile error (good); forgetting only `fireByName` is a runtime log error the new test catches.
- **Main-thread-only, no flush** — `executeQueryDynamic` asserts via `IR_ASSERT_MAIN_THREAD()`; never call from a `PARALLEL_FOR` body. Bodies making structural changes must use the deferred API; the next pipeline group boundary flushes. Zero matches is a silent no-op (correct for commands).
- **Link direction** — `IrredenEngineSystem` goes in engine/command's **PRIVATE** block so the dependency doesn't leak through the PUBLIC interface.
- **Do not route the one-shot through SystemManager** — no name/timing/concurrency slots; that's the point of "without pipeline registration". Per-call profiling, if ever wanted, is the caller's `IR_PROFILE_BLOCK`.

### Sibling + in-flight reconciliation
Open PRs checked: #2190 (rotated GRID riser shading) and #2140 (sun-shadow bake coverage) — both render-only, zero file overlap with engine/command, engine/system, or the voxel command/component headers touched here. Open-issue search: "query command" and "createSystemDynamic" return nothing open; `IRCommand`-labeled open issues = #17 only. Conflicts: none.
